flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSourceReaderMetricGroup.java

```diff
@@ -0,0 +1,141 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.metrics.groups;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.metrics.groups.OperatorIOMetricGroup;
+import org.apache.flink.metrics.groups.OperatorMetricGroup;
+import org.apache.flink.metrics.groups.SourceReaderMetricGroup;
+import org.apache.flink.metrics.groups.UnregisteredMetricsGroup;
+import org.apache.flink.runtime.metrics.MetricNames;
+import org.apache.flink.util.clock.Clock;
+import org.apache.flink.util.clock.SystemClock;
+
+/** Special {@link org.apache.flink.metrics.MetricGroup} representing an Operator. */
+@Internal
+public class InternalSourceReaderMetricGroup extends ProxyMetricGroup<MetricGroup>
+        implements SourceReaderMetricGroup {
+
+    public static final long ACTIVE = Long.MAX_VALUE;
+    private static final long UNDEFINED = Long.MIN_VALUE;
+
+    private final OperatorIOMetricGroup operatorIOMetricGroup;
+    private final Clock clock;
+    private final Counter numRecordsInErrors;
+    private long lastWatermark = UNDEFINED;
+    private long lastEventTime = UNDEFINED;
+    private long idleStartTime = ACTIVE;
+
+    private InternalSourceReaderMetricGroup(
+            MetricGroup parentMetricGroup,
+            OperatorIOMetricGroup operatorIOMetricGroup,
+            Clock clock) {
+        super(parentMetricGroup);
+        numRecordsInErrors = parentMetricGroup.counter(MetricNames.NUM_RECORDS_IN_ERRORS);
+        this.operatorIOMetricGroup = operatorIOMetricGroup;
+        this.clock = clock;
+        parentMetricGroup.gauge(
+                MetricNames.SOURCE_IDLE_TIME,
+                () -> isIdling() ? this.clock.absoluteTimeMillis() - idleStartTime : 0);
+    }
+
+    public static InternalSourceReaderMetricGroup wrap(OperatorMetricGroup operatorMetricGroup) {
+        return new InternalSourceReaderMetricGroup(
+                operatorMetricGroup,
+                operatorMetricGroup.getIOMetricGroup(),
+                SystemClock.getInstance());
+    }
+
+    @VisibleForTesting
+    public static InternalSourceReaderMetricGroup mock(MetricGroup metricGroup) {
+        return new InternalSourceReaderMetricGroup(
+                metricGroup,
+                UnregisteredMetricsGroup.createUnregisteredOperatorIOMetricGroup(),
+                SystemClock.getInstance());
+    }
+
+    private boolean isIdling() {
+        return idleStartTime != ACTIVE;
```

Comment:
It seems strange that if the source reader was idle once the idleTime will always increase. I would like to reset it somehow to 0.